crates/circuit/src/parameter/symbol_expr.rs

```diff
@@ -10,26 +10,97 @@
 // copyright notice, and modified files need to carry a notice indicating
 // that they have been altered from the originals.
 
-//! symbol_expr.rs
-//! symbolic expression engine for parameter expression
-
 use hashbrown::{HashMap, HashSet};
+use pyo3::exceptions::PyValueError;
 use std::cmp::Ordering;
 use std::cmp::PartialOrd;
 use std::convert::From;
 use std::fmt;
+use std::hash::Hash;
 use std::ops::{Add, Div, Mul, Neg, Sub};
 use std::sync::Arc;
+use uuid::Uuid;
 
 use num_complex::Complex64;
+use pyo3::prelude::*;
 
-// epsilon for SymbolExpr is hearistically defined
+// epsilon for SymbolExpr is heuristically defined
 pub const SYMEXPR_EPSILON: f64 = f64::EPSILON * 8.0;
 
+#[derive(Debug, Clone)]
+pub struct Symbol {
+    name: String,           // the name of the symbol
+    pub uuid: Uuid,         // the unique identifier
+    pub index: Option<u32>, // an optional index, if part of a vector
+    #[allow(dead_code)]
+    pub vector: Option<PyObject>, // Python only: a reference to the vector, if it is an element
+}
+
+/// Custom implementations of Eq, PartialEq, PartialOrd and Hash to ignore the ``vector`` field
+impl PartialEq for Symbol {
+    fn eq(&self, other: &Self) -> bool {
+        self.name() == other.name() && self.uuid == other.uuid && self.index == other.index
+    }
+}
+
+impl Eq for Symbol {}
+
+impl PartialOrd for Symbol {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        (&self.name, self.uuid, self.index).partial_cmp(&(&other.name, other.uuid, other.index))
+    }
+}
+
+impl Hash for Symbol {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        (&self.name, self.uuid, self.index).hash(state);
+    }
+}
+
+impl Symbol {
+    pub fn new(name: &str, uuid: Option<Uuid>, index: Option<u32>) -> Self {
+        Self {
+            name: name.to_string(),
+            uuid: uuid.unwrap_or(Uuid::new_v4()),
+            index,
+            vector: None, // Python only
+        }
+    }
+
+    /// In addition to ``new``, this also takes a vector.
+    pub fn py_new(
+        name: &str,
+        uuid: Option<u128>,
+        index: Option<u32>,
+        vector: Option<PyObject>,
+    ) -> PyResult<Self> {
+        if index.is_some() && vector.is_none() || index.is_none() && vector.is_none() {
```

Comment:
Looks like a bug: `index.is_none() && vector.is_none()`

Is there a way we could add a test to expose this?

Separately, how about:

```suggestion
        if index.is_some() != vector.is_some() {
```

Reply:
Changed to your suggestion 👍🏻 From Python I don't think we can test this since this is only ever called in `PyParameterExpression.__setstate__` and we explicitly set index and vector to `Some(..)` 🤔